src/node_http2.cc

```diff
@@ -475,25 +557,40 @@ inline void Http2Session::Settings(const nghttp2_settings_entry iv[],
 
 
 // Write data received from the i/o stream to the underlying nghttp2_session.
+// On each call to nghttp2_session_mem_recv, nghttp2 will begin calling the
+// various callback functions. Each of these will typically result in a call
+// out to JavaScript so this particular function is rather hot and can be
+// quite expensive. This is a potential performance optimization target later.
 inline ssize_t Http2Session::Write(const uv_buf_t* bufs, size_t nbufs) {
   size_t total = 0;
   // Note that nghttp2_session_mem_recv is a synchronous operation that
   // will trigger a number of other callbacks. Those will, in turn have
   // multiple side effects.
   for (size_t n = 0; n < nbufs; n++) {
+    DEBUG_HTTP2SESSION2(this, "receiving %d bytes [wants data? %d]",
+                        bufs[n].len,
+                        nghttp2_session_want_read(session_));
     ssize_t ret =
       nghttp2_session_mem_recv(session_,
                                reinterpret_cast<uint8_t*>(bufs[n].base),
                                bufs[n].len);
     CHECK_NE(ret, NGHTTP2_ERR_NOMEM);
 
-    if (ret < 0)
+    // If there is an error calling any of the callbacks, ret will be a
+    // negative number identifying the error code. This can happen, for
+    // instance, if the session is destroyed during any of the JS callbacks
+    // Note: if ssize_t is not defined (e.g. on Win32), nghttp2 will typedef
+    // ssize_t to int. Cast here so that the < 0 check actually works on
+    // Windows.
+    if (static_cast<int>(ret) < 0)
```

Comment:
While true, that does not work in this case on Windows. nghttp2 typedefs ssize_t to an int, which is incorrect and should be fixed on the nghttp2 side of things. Doing the cast the way you suggest is what I had tried first, but it causes the check to fail here.

Reply:
@jasnell We probably shouldn’t be committing `config.h` for nghttp2 here in the first place – other `#define`s like `#define SIZEOF_INT_P   4` are going to be broken on other platforms as well…

Reply:
@jasnell Actually, thinking about it *even more*, it should be fine to just `#define` it to `intptr_t` in our checked-in copy of `config.h`. That won’t make it more wrong than the define we have in `node.h`, and it will at least align the ABI between Node & nghttp2

Reply:
I would rather first see if we can upstream a fix in nghttp2 itself.

Reply:
@jasnell The wrong definition comes from `autoconf` itself… we can obviously *try* to upstream a patch into it, but I am afraid that might be too much of a breaking change.

And even if we did get it in, I think it wouldn’t even affect nghttp2’s source or direct dependencies. Rather, somebody would have to run `./configure` for nghttp2 with the new version of `autoconf` and check that file into our repository.